lib/libzpool/uzfs_mtree.c

```diff
@@ -302,3 +304,152 @@ uzfs_destroy_txg_diff_tree(void *tree)
 	avl_destroy(temp_tree);
 	umem_free(temp_tree, sizeof (*temp_tree));
 }
+
+void
+uzfs_add_to_incoming_io_tree(zvol_state_t *zv, uint64_t offset, uint64_t len)
+{
+	/*
+	 * Here: Handling of incoming_io_tree creation is for error case only.
+	 *	 It should be handled by replica or caller of uzfs_write_data
+	 */
+	if (!zv->incoming_io_tree)
+		uzfs_create_txg_diff_tree((void **)&zv->incoming_io_tree);
+
+	mutex_enter(&zv->io_tree_mtx);
+	add_to_txg_diff_tree(zv->incoming_io_tree, offset, len);
+	mutex_exit(&zv->io_tree_mtx);
+}
+
+uint32_t
+uzfs_search_incoming_io_tree(zvol_state_t *zv, uint64_t offset, uint64_t len,
+    list_t **list)
```

Comment:
I got quite confused by name of this function, thinking that search_incoming_io_tree would return an entry from the tree which overlaps with offset, len parameters. But in fact it returns all parts from the chunk from tree which are not overlapping. I'm not sure if we can find a better name. If not perhaps a comment explaining this could help.

Reply:
ok.. i will add comment :)